# 1. Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import os

In [3]:
path = r'C:\Users\bhava\Documents\18.02.26_Instacart Basket Analysis'

# Importing Data

In [4]:
ords_prods_merge_custs_viz = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_custs_viz.pkl'))

In [5]:
ords_prods_merge_custs_viz.head()

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order,product_id,add_to_cart_order,reordered,product_name,...,ord_regularity_median,frequency_flag,gender,state,age,date_joined,number_of_dependants,marital_status,income,_merge
0,2539329,1,1,2,8,NaN,196,1,0,Soda,...,20.5,Non-frequent customer,Female,Alabama,31,2/17/2019,3,married,40423,both
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,...,20.5,Non-frequent customer,Female,Alabama,31,2/17/2019,3,married,40423,both
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,...,20.5,Non-frequent customer,Female,Alabama,31,2/17/2019,3,married,40423,both
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,...,20.5,Non-frequent customer,Female,Alabama,31,2/17/2019,3,married,40423,both
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,...,20.5,Non-frequent customer,Female,Alabama,31,2/17/2019,3,married,40423,both


In [6]:
ords_prods_merge_custs_viz.tail()

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order,product_id,add_to_cart_order,reordered,product_name,...,ord_regularity_median,frequency_flag,gender,state,age,date_joined,number_of_dependants,marital_status,income,_merge
32404854,2977660,206209,13,1,12,7.0,14197,5,1,Tomato Paste,...,22.0,Non-frequent customer,Female,Iowa,74,9/14/2019,3,married,137969,both
32404855,2977660,206209,13,1,12,7.0,38730,6,0,Brownie Crunch High Protein Bar,...,22.0,Non-frequent customer,Female,Iowa,74,9/14/2019,3,married,137969,both
32404856,2977660,206209,13,1,12,7.0,31477,7,0,High Protein Bar Chunky Peanut Butter,...,22.0,Non-frequent customer,Female,Iowa,74,9/14/2019,3,married,137969,both
32404857,2977660,206209,13,1,12,7.0,6567,8,0,Chocolate Peanut Butter Protein Bar,...,22.0,Non-frequent customer,Female,Iowa,74,9/14/2019,3,married,137969,both
32404858,2977660,206209,13,1,12,7.0,22920,9,0,Roasted & Salted Shelled Pistachios,...,22.0,Non-frequent customer,Female,Iowa,74,9/14/2019,3,married,137969,both


In [7]:
ords_prods_merge_custs_viz.shape

(32404859, 31)

In [8]:
ords_prods_merge_custs_viz.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_last_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'avg_spend_user', 'spending_flag', 'ord_regularity_median',
       'frequency_flag', 'gender', 'state', 'age', 'date_joined',
       'number_of_dependants', 'marital_status', 'income', '_merge'],
      dtype='object')

In [9]:
ords_prods_merge_custs_viz.describe()

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order,product_id,add_to_cart_order,reordered,aisle_id,department_id,prices,max_order,avg_spend_user,ord_regularity_median,age,number_of_dependants,income
count,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.032876e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.239973e+07,3.240486e+07,3.240486e+07,3.240485e+07,3.240486e+07,3.240486e+07,3.240486e+07
mean,1.710745e+06,1.029372e+05,1.714230e+01,2.738867e+00,1.342515e+01,1.110408e+01,2.559866e+04,8.352547e+00,5.895873e-01,7.119612e+01,9.919792e+00,7.790994e+00,3.305217e+01,1.198023e+01,1.039776e+01,4.946527e+01,1.501896e+00,9.943773e+04
std,9.872988e+05,5.946610e+04,1.753532e+01,2.090077e+00,4.246380e+00,8.779064e+00,1.408400e+04,7.127071e+00,4.919087e-01,3.821139e+01,6.281485e+00,4.241809e+00,2.515525e+01,8.324227e+01,7.131754e+00,1.848558e+01,1.118865e+00,4.305727e+04
min,2.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,1.800000e+01,0.000000e+00,2.590300e+04
25%,8.559470e+05,5.142200e+04,5.000000e+00,1.000000e+00,1.000000e+01,5.000000e+00,1.354400e+04,3.000000e+00,0.000000e+00,3.100000e+01,4.000000e+00,4.200000e+00,1.300000e+01,7.387298e+00,6.000000e+00,3.300000e+01,1.000000e+00,6.700400e+04
50%,1.711049e+06,1.026160e+05,1.100000e+01,3.000000e+00,1.300000e+01,8.000000e+00,2.530200e+04,6.000000e+00,1.000000e+00,8.300000e+01,9.000000e+00,7.400000e+00,2.600000e+01,7.824786e+00,8.000000e+00,4.900000e+01,2.000000e+00,9.661800e+04
75%,2.565499e+06,1.543890e+05,2.400000e+01,5.000000e+00,1.600000e+01,1.500000e+01,3.794700e+04,1.100000e+01,1.000000e+00,1.070000e+02,1.600000e+01,1.130000e+01,4.700000e+01,8.254023e+00,1.300000e+01,6.500000e+01,3.000000e+00,1.279120e+05
max,3.421083e+06,2.062090e+05,9.900000e+01,6.000000e+00,2.300000e+01,3.000000e+01,4.968800e+04,1.450000e+02,1.000000e+00,1.340000e+02,2.100000e+01,2.500000e+01,9.900000e+01,2.500542e+04,3.000000e+01,8.100000e+01,3.000000e+00,5.939010e+05


# 2. PII Data

### In the previous set of Exercise 4.9 Part 2, I addressed PII info in the customer tab, dropping the name and surname of the customers to protect the identities and not trace back their details.

# 3. Creating Regional Segmentation using groups 

In [10]:
northeast = ['ME','NH','VT','MA','RI','CT','NY','PA','NJ']

In [11]:
west = ['ID','MT','WY','NV','UT','CO','AZ','NM', 'AK','WA','OR','CA','HI']

In [12]:
midwest = ['WI','MI','IL','IN','OH','ND','SD','NE','KS','MN','IA','MO']

In [13]:
south = ['DE','MD','DC','VA','WV','NC','SC','GA','FL', 'KY','TN','MS','AL', 'OK','TX','AR','LA']

In [14]:
def assign_region(state):
    if state in northeast: return 'Northeast'
    elif state in midwest: return 'Midwest'
    elif state in south: return 'South'
    elif state in west: return 'West'
    else: return 'Unknown'

In [15]:
# Mapping customers according to the region

ords_prods_merge_custs_viz['region'] = ords_prods_merge_custs_viz['state'].apply(assign_region)

In [16]:
ords_prods_merge_custs_viz['region'].value_counts()

region
Unknown    32404859
Name: count, dtype: int64

In [17]:
ords_prods_merge_custs_viz['state'].head(10)

0    Alabama
1    Alabama
2    Alabama
3    Alabama
4    Alabama
5    Alabama
6    Alabama
7    Alabama
8    Alabama
9    Alabama
Name: state, dtype: object

In [18]:
# Abbreviating state names for consistency

state_abbrev = {'Alabama':'AL','Alaska':'AK','Arizona':'AZ','Arkansas':'AR','California':'CA','Colorado':'CO', 
                'Connecticut':'CT','Delaware':'DE','District of Columbia':'DC','Florida':'FL','Georgia':'GA', 
                'Hawaii':'HI','Idaho':'ID','Illinois':'IL','Indiana':'IN','Iowa':'IA','Kansas':'KS','Kentucky':'KY',
                'Louisiana':'LA','Maine':'ME','Maryland':'MD','Massachusetts':'MA','Michigan':'MI','Minnesota':'MN', 
                'Mississippi':'MS','Missouri':'MO','Montana':'MT','Nebraska':'NE','Nevada':'NV','New Hampshire':'NH',
                'New Jersey':'NJ','New Mexico':'NM','New York':'NY','North Carolina':'NC','North Dakota':'ND',
                'Ohio': 'OH','Oklahoma':'OK','Oregon':'OR','Pennsylvania':'PA','Rhode Island':'RI',
                'South Carolina':'SC','South Dakota':'SD','Tennessee':'TN','Texas':'TX','Utah':'UT','Vermont':'VT',
                'Virginia':'VA','Washington':'WA','West Virginia':'WV','Wisconsin':'WI','Wyoming':'WY'}

In [19]:
ords_prods_merge_custs_viz['state_abbrev'] = ords_prods_merge_custs_viz['state'].map(state_abbrev)

In [20]:
ords_prods_merge_custs_viz[['state','state_abbrev']].head()

,state,state_abbrev
0,Alabama,AL
1,Alabama,AL
2,Alabama,AL
3,Alabama,AL
4,Alabama,AL


In [21]:
ords_prods_merge_custs_viz['state_abbrev'].isna().sum()

np.int64(0)

## Zero rows have failed to map a valid state abbreviation, therefore every 32,404,859 customers now has a valid state code.

# Create Census-based regional segmentation using tags Northeast, Midwest, South, West

In [22]:
ords_prods_merge_custs_viz['region'].value_counts()

region
Unknown    32404859
Name: count, dtype: int64

In [23]:
ords_prods_merge_custs_viz['region'] = ords_prods_merge_custs_viz['state_abbrev'].apply(assign_region)

In [24]:
ords_prods_merge_custs_viz['region'].value_counts()

region
South        10791885
West          8292913
Midwest       7597325
Northeast     5722736
Name: count, dtype: int64

##  As per the count, the south appears to be Instacart's largest market, while the Northeast is the smallest market. 

# 4. Excluding low activity customers and exporting out that list

In [25]:
ords_prods_merge_custs_viz['low_activity_flag'] = np.where(ords_prods_merge_custs_viz['max_order'] < 5, 1, 0)

In [26]:
ords_prods_merge_custs_viz['low_activity_flag'].value_counts()

low_activity_flag
0    30964564
1     1440295
Name: count, dtype: int64

### 30964564 rows of revenue-relevant customers and 1440295 rows of low-activity customers which are to be removed

In [27]:
ords_prods_merge_custs_viz.shape

(32404859, 34)

In [28]:
ords_prods_merge_custs_viz_cfo = ords_prods_merge_custs_viz[ords_prods_merge_custs_viz['low_activity_flag'] == 0]

In [29]:
ords_prods_merge_custs_viz_cfo.shape

(30964564, 34)

In [30]:
ords_prods_merge_custs_viz_cfo.to_pickle(os.path.join(path, "02 Data", "Prepared Data", "ords_prods_merge_custs_viz_cfo.pkl"))

### Customers with fewer than five lifetime orders have been flagged, excluded and exported out from the dataset from the CFO-level analysis. This step removes low-engagement users that generate minimal revenue and focuses all subsequent segmentation and visual analysis on economically meaningful customers.